### PR TITLE
WIP: platform, document, unit test changes of the combobox component

### PR DESCRIPTION
**Pull request by Lokanathan-k**
#### Please provide a link to the associated issue.
https://github.com/SAP/fundamental-ngx/issues/2157
#### Please provide a brief summary of this pull request.

The combobox is a selection input for selecting a value from a very large set of options. It is often connected to some backend service. It has a "typeahead" feature which enables to user to pair down the set of options based on keyword input.

Autocomplete-like component that provides Typeahead functionality needs to support variety of input types such as Array, Observable or DataSource.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done

Documentation checklist:
- [x] Documentation Examples

### CODE REVIEW

**Comment by droshev** [comment]
@Lokanathan-k since you are touching the actual library then the PR should not be prefixed as `docs`

**Comment by Lokanathan-k** [comment]
> @Lokanathan-k since you are touching the actual library then the PR should not be prefixed as `docs`

@deno there were no changes in the lib directory, we have our changes only i the document section so i have prefixed with Docs

**Comment by stefanoScalzo** [comment]
> > @Lokanathan-k since you are touching the actual library then the PR should not be prefixed as `docs`
> 
> @deno there were no changes in the lib directory, we have our changes only i the document section so i have prefixed with Docs

From what I see here it touches the libs folder...

File: libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            .compileComponents();
+    }));
+
+    beforeEach(() => {
+        fixture = TestBed.createComponent(ComboBoxComponent);
+        component = fixture.componentInstance;
+        fixture.detectChanges();
+    });
+
+    it('should create', () => {
```
Comment (KevinOkamoto) [inline libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts:39]: We should add some real unit tests here. This test is the default test and does not add to our code coverage.

Adds some test which include DataProvider and DataService and which tests that when you type a keyword in the combobox you get matching results in the combobox dropdown.

File: apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-docs.component.html [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<fd-docs-section-title [id]="'def'" [componentName]="'Combo BoX'">
```
Comment (salarenko) [inline apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-docs.component.html:1]: Combo BoX => Combo Box

File: apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-docs.component.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        },
+        {
+            language: 'CSV file',
+            code: platformComboboxAddressCSVTypesTsSrc,
+            fileName: 'addressCSV',
+            component: 'AddressCSV'
+        }
+    ];
+
+    ngOnInit() { }
```
Comment (salarenko) [inline apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-docs.component.ts:42]: Remove unused life cycle methods.

File: apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-examples/address.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+  identity(): string {
+    return this.uniqueName;
+  }
+
+  className(): string {
+    return 'Address';
+  }
+
+  toString() {
```
Comment (salarenko) [inline apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-examples/address.ts:44]: Return type

File: apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-examples/platform-combobox-types-default-example.component.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    selectedValue: string;
+
+    addressDataSource: ComboBoxDataSource<Address>;
+
+    ngOnInit() {
+        this.initDataSources();
+    }
+
+    constructor() {
```
Comment (salarenko) [inline apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-examples/platform-combobox-types-default-example.component.ts:21]: Define methods after constructor.

File: apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-examples/platform-combobox-types-default-example.component.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+  private initDataSources() {
```
Comment (salarenko) [inline apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-examples/platform-combobox-types-default-example.component.ts:24]: Return type

File: apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-header/platform-combobox-header.component.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+@Component({
+    selector: 'app-menu-header',
+    templateUrl: './platform-combobox-header.component.html',
+    styleUrls: ['./platform-combobox-header.component.scss']
+})
+export class PlatformComboboxHeaderComponent implements OnInit {
+    constructor() {}
+
+    ngOnInit() {}
```
Comment (salarenko) [inline apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-header/platform-combobox-header.component.ts:11]: Unused life cycle hook.

File: apps/docs/src/app/platform/platform-documentation.routes.ts [hunk 8]
```diff
@@ -62,6 +64,14 @@ export const ROUTES: Routes = [
                     { path: '', component: PlatformSelectDocsComponent },
                     { path: 'api', component: ApiComponent, data: { content: API_FILES.menu } }
                 ]
+            },
+            {
+                path: 'combobox',
+                component: PlatformComboboxHeaderComponent,
+                children: [
+                    { path: '', component: PlatformComboboxDocsComponent },
+                    { path: 'api', component: ApiComponent, data: { content: API_FILES.combobox } }
```
Comment (salarenko) [inline apps/docs/src/app/platform/platform-documentation.routes.ts:73]: API_FILES.combobox => API_FILES.comboBox

File: apps/docs/src/app/platform/platform-documentation.routes.ts [hunk 9]
```diff
@@ -62,6 +64,14 @@ export const ROUTES: Routes = [
                     { path: '', component: PlatformSelectDocsComponent },
                     { path: 'api', component: ApiComponent, data: { content: API_FILES.menu } }
                 ]
+            },
+            {
+                path: 'combobox',
```
Comment (salarenko) [inline apps/docs/src/app/platform/platform-documentation.routes.ts:69]: You need to be consequent, should be:
```
    path: 'combo-box',
```

File: libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+    @ViewChild(ComboBoxComponent, { static: true })
+    selectRef: ComboBoxComponent;
+
+    @ViewChild(ComboBoxComponent, { read: ElementRef, static: true })
+    selectElement: ElementRef;
+
+    wrapperValue: string;
+
+CELL_DATA: CellDataCsv[] = [
```
Comment (salarenko) [inline libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts:23]: Formatting

File: libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+export class CellData implements Entity {
+    constructor(public readonly column1: string,
+        public readonly column2: string,
+        public readonly column3: string,
+        public readonly  date: string,
+        public readonly type: string) {	
+    }
+
+    toString() {
```
Comment (salarenko) [inline libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts:49]: Return type

File: libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts (hunk 1 above)
Comment (salarenko) [reply libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts:95]: It looks like the unit tests still need some attention.

File: libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    it('component create', () => {
+        expect(component).toBeTruthy();
+    });
+
+    it('Should check for data source adding', () => {
+    component.ngOnInit();
+    component.dataSource = cellDataSource;
+    component.displayFn = toString;
+    fixture.detectChanges();
+  });
```
Comment (KevinOkamoto) [inline libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts:103]: You need to add assertions here - i.e. use `expect` function to test expected behavior.

See the following file as example:
https://github.com/SAP/fundamental-ngx/blob/master/libs/platform/src/lib/components/search-field/search-field.component.spec.ts

File: libs/platform/src/lib/domain/data-model.ts [hunk 13]
```diff
@@ -27,3 +27,30 @@ export function isSelectItem(item: SelectItem): item is SelectItem {
...
+
+/**
+ * EntityComposite having identity that can be identified in the storage by its ID. Entities are
+ * mutable objects
+ */
+export interface Entity extends Identity,
+  Deserializable,
+  CompositeType {
+}
+
```
Comment (InnaAtanasova) [inline libs/platform/src/lib/domain/data-model.ts:51]: Missing Typedocs

File: apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-examples/address.ts [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+      lines: String,
+      city: String,
+      state: String,
+      postalCode: String,
+      phone: String,
+      fax: String,
+      email: String,
+      country: String
+    };
+  }
```
Comment (InnaAtanasova) [inline apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-examples/address.ts:34]: You can create a type and return a type instead of any

Comment (Lokanathan-k) [reply apps/docs/src/app/platform/component-docs/platform-combobox/platform-combobox-examples/address.ts:34]: this method is from the parent file we are overriding the method, we cannot change the signature

File: libs/platform/src/lib/domain/data-model.ts (hunk 13 above)
Comment (Lokanathan-k) [reply libs/platform/src/lib/domain/data-model.ts:51]: > Missing Typedocs

already in place.


Screenshot 2020-04-17 at 1 52 38 PM

File: libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
```
Comment (salarenko) [inline libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts:104]: This tests definitely need to be improved to cover combo-box functionalities.

File: libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+            type: String,	
+        };
+    }
+
+    className(): string {
+        return 'CellData';
+    }
+}
+
+describe('ComboBoxComponent', () => {
```
Comment (salarenko) [inline libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts:71]: This tests definitely need to be improved to cover combo-box functionalities.

File: libs/platform/src/lib/components/form/combo-box/combo-box.component.ts [hunk 17]
```diff
@@ -89,7 +89,7 @@ export class ComboBoxComponent extends CollectionBaseInput {
     /**
      * Fill in popover values.
      */
-    _suggestions: Array<any>;
+    public _suggestions: Array<any>;
```
Comment (fkolar) [inline libs/platform/src/lib/components/form/combo-box/combo-box.component.ts:92]: why are you going this this ? `public`

File: libs/platform/src/lib/components/form/combo-box/combo-box.component.ts [hunk 18]
```diff
@@ -228,7 +229,9 @@ export class ComboBoxComponent extends CollectionBaseInput {
     private patchQueryMethod() {
         if (this._comboBox) {
             this._comboBox.handleSearchTermChange = () => {
-                this.ds.match(this._comboBox.inputTextValue);
+                if (this.ds) {
```
Comment (fkolar) [inline libs/platform/src/lib/components/form/combo-box/combo-box.component.ts:232]: Why do you need this? DS cannnot be null, if it does they there is wrong logic and we dont initialize it at certain situation.

Comment (Lokanathan-k) [reply libs/platform/src/lib/components/form/combo-box/combo-box.component.ts:232]: > Why do you need this? DS cannnot be null, if it does they there is wrong logic and we dont initialize it at certain situation.

Just a good practice to check the null, and it was fialing when datasource was passed from the unit test script files

Comment (fkolar) [reply libs/platform/src/lib/components/form/combo-box/combo-box.component.ts:232]: new typescripts has something we that you can used xx!.variable and  xxx?.variable. As I said you dont need it here. If you they are usecases where DS is null let'f fix them. With this & for this specific case you are moving some potential issue somewhere else.

File: libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+         const textInput = de.query(By.css('input.fd-input'));
+        // simulate input entry
+        textInput.nativeElement.value = 'Rowa';
+        textInput.nativeElement.dispatchEvent(new Event('input'));
+        fixture.detectChanges();
+
+        // check dropdown
+        console.log('data', host.comboBox._suggestions);
+        expect(host.comboBox._suggestions.length).toBe(3);
+    });
```
Comment (salarenko) [inline libs/platform/src/lib/components/form/combo-box/combo-box.component.spec.ts:162]: Missing tests

File: libs/platform/src/lib/components/form/combo-box/combo-box.component.ts [hunk 20]
```diff
@@ -228,7 +229,7 @@ export class ComboBoxComponent extends CollectionBaseInput {
     private patchQueryMethod() {
         if (this._comboBox) {
             this._comboBox.handleSearchTermChange = () => {
-                this.ds.match(this._comboBox.inputTextValue);
+                    this.ds.match(this._comboBox.inputTextValue);
```
Comment (droshev) [inline libs/platform/src/lib/components/form/combo-box/combo-box.component.ts:232]: this should not be part of this PR

File: libs/platform/src/lib/components/form/combo-box/combo-box.component.ts [hunk 21]
```diff
@@ -192,6 +192,7 @@ export class ComboBoxComponent extends CollectionBaseInput {
             throw new Error(`[dataSource] source did not match an array, Observable, or DataSource`);
         }
         /**
+		 *
```
Comment (droshev) [inline libs/platform/src/lib/components/form/combo-box/combo-box.component.ts:195]: this file should not be touched by the PR

**Comment by stefanoScalzo** [comment]
once an item is selected it is not allowing to reopen the combobox
Screen Shot 2020-04-27 at 2 25 13 PM.
Also, I think there is a bug with the detection of the word. If you put `US003 ` nothing comes up but `US003 - Sacramento` should come up

**Comment by droshev** [comment]
@Lokanathan-k the PR is getting old; there are conflicts and the number of commits is a bit high.
